src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -469,23 +448,19 @@ type MessageDbCategory<'event, 'state, 'context>(resolveInner, empty) =
     new (   context : MessageDbContext, codec : IEventCodec<_, _, 'context>, fold, initial,
             [<O; D(null)>]?caching,
             [<O; D(null)>]?access) =
-        let inner = Category<'event, 'state, 'context>(context, codec, ?access = access)
-        let readCacheOption =
-            match caching with
-            | None -> None
-            | Some (CachingStrategy.SlidingWindow (cache, _))
-            | Some (CachingStrategy.FixedTimeSpan (cache, _)) -> Some (cache, null)
-            | Some (CachingStrategy.SlidingWindowPrefixed (cache, _, prefix)) -> Some (cache, prefix)
-        let folder = Folder<'event, 'state, 'context>(inner, fold, initial, ?readCache = readCacheOption)
-        let category : ICategory<_, _, 'context> =
+        let raw = Category<'event, 'state, 'context>(context, codec, fold, initial, access)
+        let tryReadCache, updateCache =
+            let tryGet (cache : ICache) key = task {
+                let! cacheItem = cache.TryGet key
+                let act = Activity.Current
+                if act <> null then act.AddCacheHit(ValueOption.isSome cacheItem) |> ignore
+                return cacheItem }
             match caching with
-            | None -> folder :> _
-            | Some (CachingStrategy.SlidingWindow (cache, window)) ->
-                Caching.applyCacheUpdatesWithSlidingExpiration cache null window folder Token.supersedes
-            | Some (CachingStrategy.FixedTimeSpan (cache, period)) ->
-                Caching.applyCacheUpdatesWithFixedTimeSpan cache null period folder Token.supersedes
-            | Some (CachingStrategy.SlidingWindowPrefixed (cache, window, prefix)) ->
-                Caching.applyCacheUpdatesWithSlidingExpiration cache prefix window folder Token.supersedes
-        let resolveInner categoryName streamId = struct (category, StreamName.render categoryName streamId, ValueNone)
+            | None -> (fun _ -> Task.FromResult ValueNone), fun _ _ -> Task.FromResult ()
+            | Some (CachingStrategy.SlidingWindow (cache, window)) -> tryGet cache, Cache.updateWithSlidingExpiration (cache, null) window Token.supersedes
+            | Some (CachingStrategy.FixedTimeSpan (cache, period)) -> tryGet cache, Cache.updateWithFixedTimeSpan (cache, null) period Token.supersedes
+            | Some (CachingStrategy.SlidingWindowPrefixed (cache, window, prefix)) -> tryGet cache, Cache.updateWithSlidingExpiration (cache, prefix) window Token.supersedes
+        let cached = CachingDecorator<'event, 'state, 'context>(raw, tryReadCache, updateCache) : ICategory<_, _, _>
```

Comment:
slight difference in behaviour vs what was here previously. Previously we'd not instantiate the wrapper if no caching was desired.

Not the biggest of deals, a couple of wasted cycles per call, just thought I'd call it out

Reply:
Hm yeah it should be possible to restore the previous logic by wrapping the instantiation of the Decorator in a small helper and having the preceding match yield the decorated form or the raw one depending on whether it's needed

Reply:
null wrapping now removed in Caching.fs